### PR TITLE
feat: scheduler fn support

### PR DESCRIPTION
**Pull request by Salakar**
## Testing

Deployed the following sample:

```python
from firebase_functions import scheduler_fn


@scheduler_fn.on_schedule(
    schedule="* * * * *",
    timezone=scheduler_fn.Timezone("America/Los_Angeles"),
)
def example(event: scheduler_fn.ScheduledEvent) -> None:
    print("Hello from scheduled function!")
    print(event.job_name)
    print(event.schedule_time)
```

**Deployment is successful:**

image

image

**Schedule runs and prints:**
image

### CODE REVIEW

File: src/firebase_functions/scheduler_fn.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    """
+
+    job_name: str | None
+    """
+    The Cloud Scheduler job name.
+    Populated via the X-CloudScheduler-JobName header.
+    If invoked manually, this field is None.
+    """
+
+    schedule_time: _dt.datetime
```
Comment (taeold) [inline src/firebase_functions/scheduler_fn.py:48]: nit* This is great! I'd been fine with string, but parsing it sounds extra helpful!

File: tests/test_scheduler_fn.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        Tests to ensure the decorated function is called correctly
+        with appropriate ScheduledEvent object and returns a 200
+        status code if successful.
+        """
+
+        with Flask(__name__).test_request_context("/"):
+            environ = EnvironBuilder(
+                headers={
+                    "X-CloudScheduler-JobName": "example-job",
+                    "X-CloudScheduler-ScheduleTime": "2023-04-13T12:00:00-07:00"
```
Comment (taeold) [inline tests/test_scheduler_fn.py:58]: nit* Thin we are missing a `Z` at the end.

File: src/firebase_functions/scheduler_fn.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        def on_schedule_wrapped(request: _Request) -> _Response:
+            schedule_time: _dt.datetime
+            schedule_time_str = request.headers.get(
+                "X-CloudScheduler-ScheduleTime")
+            if schedule_time_str is None:
+                schedule_time = _dt.datetime.utcnow()
+            else:
+                schedule_time = _dt.datetime.strptime(
+                    schedule_time_str,
+                    "%Y-%m-%dT%H:%M:%S%z",
```
Comment (taeold) [inline src/firebase_functions/scheduler_fn.py:104]: I think this is missing a trailing "Z".

IIUC, `datetime.fromisoformat()` could also work here?

Comment (taeold) [reply src/firebase_functions/scheduler_fn.py:104]: But then again - you tested the code in prod 🤔 wonder how it worked.

**Review by taeold — APPROVED** [review]
minor nits

File: src/firebase_functions/scheduler_fn.py (hunk 3 above)
Comment (Salakar) [reply src/firebase_functions/scheduler_fn.py:104]: Original code expected the Z in the test date time, in prod the dates were coming through without this and weren't an ISO format so this different parsing was needed; here's a log from prod from previous implementation showing the format: 

```
ValueError: time data '2023-04-13T10:19:00-07:00' does not match format '%Y-%m-%dT%H:%M:%SZ'

at ._strptime ( /layers/google.python.runtime/python/lib/python3.10/_strptime.py:349 )
at ._strptime_datetime ( /layers/google.python.runtime/python/lib/python3.10/_strptime.py:568 )
at .on_schedule_wrapped ( /layers/google.python.pip/pip/lib/python3.10/site-packages/firebase_functions/scheduler_fn.py:102 )
at .view_func ( /layers/google.python.pip/pip/lib/python3.10/site-packages/functions_framework/__init__.py:130 )
```

File: tests/test_scheduler_fn.py (hunk 2 above)
Comment (Salakar) [reply tests/test_scheduler_fn.py:58]: Original code expected the Z in the test date time, in prod the dates were coming through without this and weren't an ISO format so this different parsing was needed; here's a log from prod from previous implementation showing the format: 

```
ValueError: time data '2023-04-13T10:19:00-07:00' does not match format '%Y-%m-%dT%H:%M:%SZ'

at ._strptime ( /layers/google.python.runtime/python/lib/python3.10/_strptime.py:349 )
at ._strptime_datetime ( /layers/google.python.runtime/python/lib/python3.10/_strptime.py:568 )
at .on_schedule_wrapped ( /layers/google.python.pip/pip/lib/python3.10/site-packages/firebase_functions/scheduler_fn.py:102 )
at .view_func ( /layers/google.python.pip/pip/lib/python3.10/site-packages/functions_framework/__init__.py:130 )
```

I updated the tests after to match the same format as prod

File: src/firebase_functions/scheduler_fn.py (hunk 3 above)
Comment (taeold) [reply src/firebase_functions/scheduler_fn.py:104]: That's good to know. Thanks for double checking!

**Pull request merged**